Added search feature to tags index to search all tags

dscottS3 opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/main/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Add a text field to top tags page which allows searching of all tags

## Related Tickets & Documents

#15150 

## QA Instructions, Screenshots, Recordings

Visit `/tags` and search for a tag. You can also visit `/tags?q=sometag`.

### UI accessibility concerns?

There shouldn't be. It uses the same search form that's in the header of the site

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

citizen428 commented on app/controllers/tags_controller.rb:
```diff
@@ -18,7 +18,11 @@ class TagsController < ApplicationController
     skip_authorization
     @tags_index = true
-    @tags = Tag.direct.includes(:sponsorship).order(hotness_score: :desc).limit(100)
+    @tags = if params.key?(:q) && params[:q].present?
```
Suggestion: This seems almost too defensive, the `.present?` check alone seems to cover all relevant cases, doesn't it?

![Screen Shot 2022-01-25 at 11 13 48](https://user-images.githubusercontent.com/47985/150909722-834a7b9d-b766-481b-a3bd-92bd501ef31e.png)

```suggestion
    @tags = if params[:q].present?
```

citizen428 commented on app/controllers/tags_controller.rb:
```diff
@@ -18,7 +18,11 @@ class TagsController < ApplicationController
     @tags_index = true
-    @tags = Tag.direct.includes(:sponsorship).order(hotness_score: :desc).limit(100)
+    @tags = if params.key?(:q) && params[:q].present?
+              Tag.includes([:sponsorship]).search_by_name(params[:q]).limit(100)
```
Question: should we move the LIMIT into a constant so that we can keep it consistent between both branches without having to make two changes?

jeremyf replied:
Another option would be:

```ruby
@tags = Tag.includes([:sponsorship]).limit(100)
if params[:q].present?
  @tags = @tags.search_by_name(params[:q])
else
  @tags = @tags.order(hotness_score: :desc)
end
```

Throughout the code-base we have several places that uses and replaces to an initial scope.

dscottS3 replied:
I think I like that @jeremyf's suggestion a tad better. in fact it's what I had initially but as `before_action` that initially defined and memoized `@tags` for the index action. I may throw up a commit with this just to collect feedback.

citizen428 commented on app/views/tags/index.html.erb:
```diff
@@ -21,8 +21,23 @@
-    <h1 class="crayons-title"><%= t("views.tags.heading") %></h1>
-    <%= link_to t("views.tags.following_tags"), dashboard_following_tags_path, class: "crayons-btn crayons-btn--ghost" %>
+    <h1 class="crayons-title">
+      <% if params.key?(:q) && params[:q].present? %>
```
Suggestion: same as above.

```suggestion
      <% if params[:q].present? %>
```

dscottS3 replied:
thanks for these suggestions! I feel like I've been bitten in the past by not explicitly checking if the key was in the `params`. maybe that was way back in some older version of rails/ruby or a wonky codebase :she

citizen428 replied:
You definitely can get bitten by that but generally not when there's a presence check, as they are somewhat redundant. It definitely can happen with hashes and boolean values though, since `false` is not `present?`. 

As you said, things changed a lot over the years, we have the safe-navigation operator now (`&.`), etc. so this should be good 😃

citizen428 commented on app/controllers/tags_controller.rb:
```diff
@@ -61,6 +61,10 @@ def suggest
   private
 
+  def tags
+    @tags ||= Tag.direct.includes(:sponsorship).limit 100
```
Interesting that Rubocop didn't flag this. We basically reserve method calls without parens for DSL-like class "macros", like `belongs_to`, etc. Not a big deal, but seeing two different styles used on the same line sends me on an OCD spiral 😆 

```suggestion
    @tags ||= Tag.direct.includes(:sponsorship).limit(100)
```

dscottS3 replied:
haha makes sense! No worries, It's admittedly a habit of mine. Whether it's good or bad is debatable 😆

aitchiss commented on app/views/tags/index.html.erb:
```diff
@@ -21,8 +21,23 @@
+    <form accept-charset="UTF-8" method="get" action="<%= tags_path %>" role="search">
+      <div class="crayons-fields crayons-fields--horizontal">
+        <div class="crayons-field flex-1 relative">
+          <input class="crayons-header--search-input crayons-textfield js-search-input" type="text" id="nav-search" name="q" placeholder="<%= t("views.tags.search.placeholder") %>" autocomplete="off" />
```
Just a small nitpick - since we don't have a visible label on the page, we should add an `aria-label` so screen reader users know what the input is for. I think it could just be the same as the placeholder, but maybe it's cleaner to create a separate i18n key for it

dscottS3 replied:
added!

aitchiss commented on app/views/tags/index.html.erb:
```diff
@@ -21,8 +21,23 @@
+      <div class="crayons-fields crayons-fields--horizontal">
+        <div class="crayons-field flex-1 relative">
+          <input class="crayons-header--search-input crayons-textfield js-search-input" type="text" id="nav-search" name="q" placeholder="<%= t("views.tags.search.placeholder") %>" autocomplete="off" />
+          <button type="submit" aria-label="<%= t("views.search.icon.aria_label") %>" class="c-btn c-btn--icon-alone absolute inset-px left-auto mt-0 py-0">
```
Another wee nitpick - it would be great if we could create a separate aria-label i18n string for this, otherwise screen reader users can't differentiate easily between the general search and the tags search. Something like "search tags" would work well

dscottS3 replied:
added here too! 😄

aitchiss reviewed (commented):
This is working really well! I added a couple of small accessibility requests, but one other thing I noticed is that we don't have an empty state for when no search results are found - e.g.

![tags view with empty results](https://user-images.githubusercontent.com/20773163/151172398-64b9fb53-cc15-40ad-ba49-3ab06a3143d6.png)

We should probably display something like we do on the main search page ("No results match that query").

It would also be cool to add some Cypress tests for this new user flow, but I don't mind if that's a follow up PR (I'd be happy to add them once this is merged, if you like?)

citizen428 commented:
@dscottS3 Since this was the first new PR I noticed that had a review suggestion from the new Rubocop GH action I added recently, I took the liberty of accepting it which means this feature has now been tested end to end. Hope that's ok 😃

citizen428 commented on spec/requests/tags_spec.rb:
```diff
@@ -17,6 +17,18 @@
     end
+
+    it "searches tags" do
+      %w[ruby java javascript).each { |t| create(:tag, name: t) }
```
Wow, the Rubocop auto fix messed this up by only changing the opening `(` to `[`, sorry for that. I'm gonna fix it.

dscottS3 replied:
haha cool. I seen that and seen that it didn't have the closing bracket. I fixed it locally but haven't pushed it up, yet.

dscottS3 commented:
@aitchiss I've updated the empty state to be more like the main search. also, I've added a request spec that checks the body for `No results match that query` if the search result is empty. not sure if that is satisfactory enough for you in regards to a cypress test.

citizen428 commented on spec/requests/tags_spec.rb:
```diff
@@ -17,6 +17,21 @@
       expect(response.body).not_to include("aliastag")
     end
+
+    it "searches tags" do
```
In general, we add the [aggregate_failures](https://relishapp.com/rspec/rspec-core/docs/expectation-framework-integration/aggregating-failures) metadata on specs with more than one `expect`:

```suggestion
    it "searches tags", :aggregate_failures do
```

aitchiss reviewed (approved):
Looks good to me - thanks for making those tweaks!
